src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/ProjectState.cs

```diff
@@ -274,26 +274,36 @@ public ProjectState WithChangedHostDocument(HostDocument hostDocument, TextLoade
 
     public ProjectState WithHostProjectAndWorkspaceState(HostProject hostProject, ProjectWorkspaceState projectWorkspaceState)
     {
-        if (HostProject.Configuration.Equals(hostProject.Configuration) &&
-            HostProject.RootNamespace == hostProject.RootNamespace &&
-            ProjectWorkspaceState.Equals(projectWorkspaceState))
+        var configUnchanged = (HostProject.Configuration.Equals(hostProject.Configuration) &&
+            HostProject.RootNamespace == hostProject.RootNamespace);
+
+        if (configUnchanged && ProjectWorkspaceState.Equals(projectWorkspaceState))
         {
             return this;
         }
 
-        var documents = Documents.ToImmutableDictionary(kvp => kvp.Key, kvp => kvp.Value.WithProjectChange(), FilePathNormalizingComparer.Instance);
+        var documents = Documents.ToImmutableDictionary(kvp => kvp.Key, kvp => kvp.Value.WithProjectChange(cacheComputedState: configUnchanged), FilePathNormalizingComparer.Instance);
+
+        var importsToRelatedDocuments = configUnchanged
+            ? ImportsToRelatedDocuments
+            : ComputeImportsToRelatedDocuments(documents);
 
-        // If the host project has changed then we need to recompute the imports map
-        var importsToRelatedDocuments = s_emptyImportsToRelatedDocuments;
+        var state = new ProjectState(this, numberOfDocumentsMayHaveChanged: !configUnchanged, hostProject, projectWorkspaceState, documents, importsToRelatedDocuments);
+        return state;
 
-        foreach (var document in documents)
+        ImmutableDictionary<string, ImmutableArray<string>> ComputeImportsToRelatedDocuments(ImmutableDictionary<string, DocumentState> documents)
         {
-            var importTargetPaths = GetImportDocumentTargetPaths(document.Value.HostDocument);
-            importsToRelatedDocuments = AddToImportsToRelatedDocuments(importsToRelatedDocuments, document.Value.HostDocument.FilePath, importTargetPaths);
-        }
+            // If the host project has changed then we need to recompute the imports map
```

Comment:
nit: this comment is probably not needed since we're lazily computing. Or put the comment where the call to this is

Reply:
Good call, thanks